### ISSUE TITLE
Strange "object does not exist error" for SWI-Prolog 7

### ISSUE THREAD

**Issue by dram** [LogtalkDotOrg/logtalk3#51]
For following code, I got a strange `object 'http_server' does not exist` error in SWI-Prolog 7.5.2-16-gdc20d77 (also in 7.2.3), but it's ok in SWI 6.6.6.

```
% cat sources/loader.lgt
:- use_module(library(http/thread_httpd), []).

:- use_module(library(http/http_dispatch), []).

:- initialization((
        logtalk_load(http_server)
)).
```

```
% cat sources/http_server.lgt
:- object(http_server).

        :- public(run/0).

        run :-
                thread_httpd:http_server(
                        http_dispatch:http_dispatch,
                        [port('0.0.0.0':8000)]).

:- end_object.
```

```
% cat run.sh
#!/bin/sh

ROOT="$(dirname "$(readlink -f "$0")")"

export LOGTALKHOME="$ROOT/../logtalk3"
export LOGTALKUSER="$ROOT/../logtalk3"

GOAL="{'$ROOT/sources/loader.lgt'}, http_server::run"

export PATH="$HOME/prolog/swi/master/bin/:$PATH"

exec "$LOGTALKHOME"/integration/swilgt.sh -q -g "$GOAL"
```

```
% ./run.sh
Warning: /home/dram/logtalk/http-server/sources/http_server.lgt:10:
        /home/dram/logtalk/http-server/sources/.lgt_tmp/loader_10247014_lgt.pl:6: Initialization goal failed
ERROR: /home/dram/logtalk/http-server/sources/http_server.lgt:1:
        -g {'/home/dram/logtalk/http-server/sources/loader.lgt'}, http_server::run: object `http_server' does not exist
```

**Comment by dram** [comment]
Sorry, forgot to mention Logtalk version, I'm using master branch, Logtalk 3.10.4-rc1.

**Comment by pmoura** [comment]
I can reproduce the loading failure of the `http_server` object. It seems to be an issue with the compilation of the qualified modules calls in the body of the `run/0` predicate. That the code works on SWI-Prolog 6.6.6 but not in 7.5.2 suggest a change in the meta-predicate templates of the HTTP module predicates. I'm looking into it. As a temporary workaround, use the `{}/1` compiler bypass control construct in the definition of the `run/0` predicate:

```logtalk
        run :-
                {thread_httpd:http_server(
                        http_dispatch:http_dispatch,
                        [port('0.0.0.0':8000)])}.
```

**Comment by pmoura** [comment]
In SWI-Prolog 7.5.2 we get:

```logtalk
?- predicate_property(thread_httpd:http_server(_,_), meta_predicate(M)).
M = http_server(1, :).
```

The second meta-argument in this meta-predicate template is ambiguous (`:`). Thus, Logtalk doesn't have enough information to compile the argument. But the compiler is expected to print an error in this case and not just fail to compile the call silently. That's definitely a bug.

**Comment by pmoura** [comment]
Fixed in 79dacb08758f3b19dccaeda0914815cc00e1d7cb. Thanks for the bug report. The compiler nows reports the expected error. The best solution is to use `{}/1` control construct as explained above. An alternative solution would be to override the meta-predicate template for the module predicate.

**Comment by dram** [comment]
Fix confirmed, thanks!
